### PR TITLE
Improved NetFlowV9 support

### LINKED ISSUE #18: Netflow v9 Flows with no template present (yet) should be stored instead of discarded.
Right now we are [discarding flows](https://github.com/graylog-labs/graylog-plugin-netflow/commit/f00cf728e012786069cef46b60848ffe64e8309d) when the associated template is not present (yet), while [RFC3954](https://www.ietf.org/rfc/rfc3954.txt) states that:

```
   If the Template Records have not been
   received at the time Flow Data Records (or Options Data Records) are
   received, the Collector SHOULD store the Flow Data Records (or
   Options Data Records) and decode them after the Template Records are
   received.  A Collector device MUST NOT assume that the Data FlowSet
   and the associated Template FlowSet (or Options Template FlowSet) are
   exported in the same Export Packet.
```

Temporary buffering of flows where the template is not present (yet) should be implemented and once the template was received, associated flows should be processed.

### LINKED ISSUE #19: v9 template parsing issue
There seems to be a problem with template parsing. The first error shows that there is no template for a packet. The second error, 2 seconds later, shows a template parsing issue.

```
2017-08-08_13:58:03.28568 2017-08-08 13:58:03,285 ERROR: org.graylog.plugins.netflow.codecs.NetFlowCodec - Error parsing NetFlow packet <b78b2476-7c40-11e7-a3ff-005056b6418d> received from <10.1.10.26:54482>
2017-08-08_13:58:03.28569 org.graylog.plugins.netflow.flows.EmptyTemplateException: Unable to parse NetFlow 9 records without template. Discarding packet.
2017-08-08_13:58:03.28569 	at org.graylog.plugins.netflow.v9.NetFlowV9Parser.parsePacket(NetFlowV9Parser.java:56) ~[graylog-plugin-netflow-2.3.0-rc.4.jar:?]
2017-08-08_13:58:03.28569 	at org.graylog.plugins.netflow.flows.NetFlowParser.parse(NetFlowParser.java:63) ~[graylog-plugin-netflow-2.3.0-rc.4.jar:?]
2017-08-08_13:58:03.28569 	at org.graylog.plugins.netflow.codecs.NetFlowCodec.decodeMessages(NetFlowCodec.java:107) [graylog-plugin-netflow-2.3.0-rc.4.jar:?]
2017-08-08_13:58:03.28569 	at org.graylog2.shared.buffers.processors.DecodingProcessor.processMessage(DecodingProcessor.java:144) [graylog.jar:?]
2017-08-08_13:58:03.28570 	at org.graylog2.shared.buffers.processors.DecodingProcessor.onEvent(DecodingProcessor.java:87) [graylog.jar:?]
2017-08-08_13:58:03.28570 	at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:74) [graylog.jar:?]
2017-08-08_13:58:03.28570 	at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:42) [graylog.jar:?]
2017-08-08_13:58:03.28570 	at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
2017-08-08_13:58:03.28570 	at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
2017-08-08_13:58:03.28571 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]

2017-08-08_13:58:05.34156 2017-08-08 13:58:05,341 ERROR: org.graylog2.shared.buffers.processors.DecodingProcessor - Unable to decode raw message RawMessage{id=b81c8f00-7c40-11e7-a3ff-005056b6418d, journalOffset=250337669, codec=netflow, payloadSize=492, timestamp=2017-08-08T13:51:46.672Z, remoteAddress=/10.1.9.2:7560} on input <5980f99075010f0b154d87e9>.
2017-08-08_13:58:05.34220 2017-08-08 13:58:05,341 ERROR: org.graylog2.shared.buffers.processors.DecodingProcessor - Error processing message RawMessage{id=b81c8f00-7c40-11e7-a3ff-005056b6418d, journalOffset=250337669, codec=netflow, payloadSize=492, timestamp=2017-08-08T13:51:46.672Z, remoteAddress=/10.1.9.2:7560}
2017-08-08_13:58:05.34221 java.lang.IndexOutOfBoundsException: readerIndex(492) + length(2) exceeds writerIndex(492): UnpooledHeapByteBuf(ridx: 492, widx: 492, cap: 492/492)
2017-08-08_13:58:05.34222 	at io.netty.buffer.AbstractByteBuf.checkReadableBytes0(AbstractByteBuf.java:1395) ~[graylog.jar:?]
2017-08-08_13:58:05.34223 	at io.netty.buffer.AbstractByteBuf.readShort(AbstractByteBuf.java:706) ~[graylog.jar:?]
2017-08-08_13:58:05.34223 	at io.netty.buffer.AbstractByteBuf.readUnsignedShort(AbstractByteBuf.java:722) ~[graylog.jar:?]
2017-08-08_13:58:05.34224 	at org.graylog.plugins.netflow.v9.NetFlowV9Parser.parseTemplates(NetFlowV9Parser.java:122) ~[?:?]
2017-08-08_13:58:05.34224 	at org.graylog.plugins.netflow.v9.NetFlowV9Parser.parsePacket(NetFlowV9Parser.java:45) ~[?:?]
2017-08-08_13:58:05.34224 	at org.graylog.plugins.netflow.flows.NetFlowParser.parse(NetFlowParser.java:63) ~[?:?]
2017-08-08_13:58:05.34224 	at org.graylog.plugins.netflow.codecs.NetFlowCodec.decodeMessages(NetFlowCodec.java:107) ~[?:?]
2017-08-08_13:58:05.34224 	at org.graylog2.shared.buffers.processors.DecodingProcessor.processMessage(DecodingProcessor.java:144) ~[graylog.jar:?]
2017-08-08_13:58:05.34225 	at org.graylog2.shared.buffers.processors.DecodingProcessor.onEvent(DecodingProcessor.java:87) [graylog.jar:?]
2017-08-08_13:58:05.34225 	at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:74) [graylog.jar:?]
2017-08-08_13:58:05.34225 	at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:42) [graylog.jar:?]
2017-08-08_13:58:05.34225 	at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
2017-08-08_13:58:05.34225 	at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
2017-08-08_13:58:05.34225 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
```

### LINKED ISSUE #20: Unexpected end-of-input in field name
When I configure Palo Alto to send netflow packets, this error occurs:

`2017-08-18T18:20:18.938-03:00 ERROR [NetFlowV9TemplateCache] Couldn't load template cache from disk
com.fasterxml.jackson.core.io.JsonEOFException: Unexpected end-of-input in field name
 at [Source: /tmp/netflow-templates.json; line: 1, column: 16001]
        at com.fasterxml.jackson.core.base.ParserMinimalBase._reportInvalidEOF(ParserMinimalBase.java:483) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.parseEscapedName(UTF8StreamJsonParser.java:2020) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.slowParseName(UTF8StreamJsonParser.java:1921) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._parseName(UTF8StreamJsonParser.java:1705) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:776) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:144) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBind(MapDeserializer.java:458) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:365) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:27) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3814) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2775) ~[graylog.jar:?]
        at org.graylog.plugins.netflow.v9.NetFlowV9TemplateCache.loadCache(NetFlowV9TemplateCache.java:71) [graylog-plugin-netflow-2.3.0-rc.4.jar:?]
        at org.graylog.plugins.netflow.v9.NetFlowV9TemplateCache.<init>(NetFlowV9TemplateCache.java:62) [graylog-plugin-netflow-2.3.0-rc.4.jar:?]
        at org.graylog.plugins.netflow.v9.NetFlowV9TemplateCache.<init>(NetFlowV9TemplateCache.java:48) [graylog-plugin-netflow-2.3.0-rc.4.jar:?]
        at org.graylog.plugins.netflow.codecs.NetFlowCodec.<init>(NetFlowCodec.java:85) [graylog-plugin-netflow-2.3.0-rc.4.jar:?]
        at org.graylog.plugins.netflow.codecs.NetFlowCodec$$FastClassByGuice$$5770acd4.newInstance(<generated>) [graylog-plugin-netflow-2.3.0-rc.4.jar:?]
        at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:89) [graylog.jar:?]
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:111) [graylog.jar:?]
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:90) [graylog.jar:?]
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:268) [graylog.jar:?]
        at com.google.inject.internal.InjectorImpl$2$1.call(InjectorImpl.java:1019) [graylog.jar:?]
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1085) [graylog.jar:?]
        at com.google.inject.internal.InjectorImpl$2.get(InjectorImpl.java:1015) [graylog.jar:?]
        at com.google.inject.assistedinject.FactoryProvider2.invoke(FactoryProvider2.java:776) [graylog.jar:?]
        at com.sun.proxy.$Proxy118.create(Unknown Source) [?:?]
        at org.graylog2.shared.buffers.processors.DecodingProcessor.processMessage(DecodingProcessor.java:132) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.DecodingProcessor.onEvent(DecodingProcessor.java:87) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:74) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:42) [graylog.jar:?]
        at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]`

Below is the template received

[netflow-templates.txt](https://github.com/graylog-labs/graylog-plugin-netflow/files/1235718/netflow-templates.txt)

### PR DESCRIPTION
**Pull request by kroepke**
This PR changes the way templates are handled.

Since for V9 template data flows are not sent with every packet, the implementation must buffer packets until it receives the necessary templates to know how to parse them. The same is true for the option template.

This implementation moves the buffering and template aggregation into a custom codec aggregator, so that the codec itself, which runs after journalling the message, can assume that it has all the templates it needs to successfully parse a packet. This is even more important when processing a journal after a restart.
The RFC requires not to write templates to disk (or otherwise store them independent of data flows), because they might change at any time. We therefore colocate them with the data itself, taking the performance hit of writing more bytes to disk, but with the benefit of a safer implementation.

Thus this implementation does not lose data in the case it doesn't have templates yet. Those are resent regularly by the exporter, for each observation domain.

To be compatible with Graylog 2.3, this change comes with a custom codec aggregator, for 3.0 we can migrate the code back into the server.

fixes #18 
fixes #19 
fixes #20

### CODE REVIEW

File: src/main/java/org/graylog/plugins/netflow/codecs/NetflowV9CodecAggregator.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            resultBuffer.writeBytes(builder.build().toByteArray());
+            return new Result(resultBuffer, true);
+
+        } catch (Exception e) {
+            LOG.error("Unexpected failure while aggregating NetFlowV9 packet, discarding packet.", ExceptionUtils.getRootCause(e));
+            return new Result(null, false);
+        }
+    }
+
+    private void queueBufferedPackets(Set<TemplateKey> templates, Set<ChannelBuffer> packetsToSend, TemplateKey templateKey) {
```
Comment (kroepke) [inline src/main/java/org/graylog/plugins/netflow/codecs/NetflowV9CodecAggregator.java]: This actually requires checking all templates used by a packet, so this code will change soon and is not the final version.

**Comment by kroepke** [comment]
Ah shit. Yeah the prefix is broken now.
I'll fix it tomorrow

**Comment by bernd** [comment]
I also saw this. Can this still happen? I thought we are waiting until we get a template before we pass on the packet.

```
2017-08-24 20:24:01,298 ERROR: org.graylog.plugins.netflow.codecs.NetFlowCodec - Error parsing NetFlow packet <66e509d0-88f9-11e7-8379-024278d7ba18> received from <127.0.0.1:42934>
org.graylog.plugins.netflow.flows.EmptyTemplateException: Unable to parse NetFlow 9 records without template. Discarding packet.
	at org.graylog.plugins.netflow.v9.NetFlowV9Parser.parsePacket(NetFlowV9Parser.java:65) ~[classes/:?]
	at org.graylog.plugins.netflow.codecs.NetFlowCodec.lambda$decodeV9Packets$7(NetFlowCodec.java:186) ~[classes/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:1.8.0_144]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:1.8.0_144]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374) ~[?:1.8.0_144]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_144]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_144]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:1.8.0_144]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_144]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[?:1.8.0_144]
	at org.graylog.plugins.netflow.codecs.NetFlowCodec.decodeV9Packets(NetFlowCodec.java:187) ~[classes/:?]
	at org.graylog.plugins.netflow.codecs.NetFlowCodec.decodeV9(NetFlowCodec.java:156) ~[classes/:?]
	at org.graylog.plugins.netflow.codecs.NetFlowCodec.decodeMessages(NetFlowCodec.java:134) [classes/:?]
	at org.graylog2.shared.buffers.processors.DecodingProcessor.processMessage(DecodingProcessor.java:144) [classes/:?]
	at org.graylog2.shared.buffers.processors.DecodingProcessor.onEvent(DecodingProcessor.java:87) [classes/:?]
	at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:74) [classes/:?]
	at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:42) [classes/:?]
	at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [disruptor-3.3.6.jar:?]
	at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [metrics-core-3.2.2.jar:3.2.2]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
```

**Comment by kroepke** [comment]
That's weird. Do you have a config that causes this?
If nothing else I'll try to get better logging in there (e.g. printing the packet structure for when that error happens).

**Comment by kroepke** [comment]
The field 153 is because our default field definition list is missing the type. It is "flow end milliseconds" of https://www.iana.org/assignments/ipfix/ipfix.xhtml
We should probably ship with a better default list, taking the one from IPFIX.

**Comment by joschi** [comment]
> We should probably ship with a better default list, taking the one from IPFIX.

I would strongly advise against this, as IPFIX ("NetFlow version 10") has some incompatible fields with NetFlow version 9.

For example, id 1 is "octetDeltaCount" in IPFIX, while it's the number of incoming bytes in NetFlow 9 (http://netflow.caligare.com/netflow_v9.htm).

**Comment by kroepke** [comment]
@joschi Fair point, the names sometimes differ, but semantically IPFIX is the same as NetFlow v9 when it comes to the 127 fields v9 defined (as stated in [RFC 5102](https://tools.ietf.org/html/rfc5102#section-4))

At the very least the definition list we ship should be consistent in the IPFIX extensions, while retaining the original v9 fields.

**Comment by michelealbrigo** [comment]
I can confirm that #20 looks solved now: I am running rc5 and it's dealing with our full netflow configuration without any apparent problem since some hours.

**Comment by moadiv** [comment]
Hi everybody, i'm having troubles with an Invalid FlowVersion Exception (Invalid NetFlow version 0) when trying to log the flow of a netgear switch.
This is my environment:

Graylog v2.3.1+9f2c6ef
graylog-plugin-netflow-2.3.0-rc.5.jar
switch netgear M5300-28G3 sflow V5

And this is the server.log result:

2017-09-05T12:02:13.080-04:00 ERROR [NetFlowCodec] Error parsing NetFlow packet <94943050-9253-11e7-9ea1-000c29f4e228> received from <192.168.128.254:6343>
org.graylog.plugins.netflow.flows.InvalidFlowVersionException: **Invalid NetFlow version 0**
        at org.graylog.plugins.netflow.v5.NetFlowV5Parser.parseHeader(NetFlowV5Parser.java:67) ~[graylog-plugin-netflow-2.3.0-rc.5.tmp:?]
        at org.graylog.plugins.netflow.v5.NetFlowV5Parser.parsePacket(NetFlowV5Parser.java:33) ~[graylog-plugin-netflow-2.3.0-rc.5.tmp:?]
        at org.graylog.plugins.netflow.codecs.NetFlowCodec.decodeMessages(NetFlowCodec.java:127) [graylog-plugin-netflow-2.3.0-rc.5.tmp:?]
        at org.graylog2.shared.buffers.processors.DecodingProcessor.processMessage(DecodingProcessor.java:144) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.DecodingProcessor.onEvent(DecodingProcessor.java:87) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:74) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:42) [graylog.jar:?]
        at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:143) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_141]

Thanks in advance !!!!

**Comment by joschi** [comment]
@moadiv This plugin currently doesn't support SFlow, see #3.

Please questions about this plugin to our [discussion forum](https://community.graylog.org/) or join the [#graylog channel on freenode IRC](https://webchat.freenode.net/?channels=%23graylog).

Thank you!

**Pull request merged**